### PR TITLE
[REEF-826] Add state transition check in ResourceManagerStatus.setState

### PR DESCRIPTION
**Pull request by zerg-zerg**
This patch:
 * Adds the isLegalStateTransition() method to ResourceManagerStatus
 * Checks the state transition in the method according to the transition diagram from REEF-826
 * Calls the method in setState()
 * Throws IllegalStateException if the state transition is prohibited

JIRA:
 [REEF-826] (https://issues.apache.org/jira/browse/REEF-826)

### CODE REVIEW

**Comment by afchung** [comment]
I'll do a pass.

**Comment by afchung** [comment]
This looks good to me. Reading through the JIRA discussion, it seems like you and @markusweimer still have some unresolved issues with regards to transition of the ``SUSPEND`` value, so I'll wait for him to sign off as well to test and merge. Thanks!

**Comment by zerg-zerg** [comment]
@markusweimer Could you please have a look at the request?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceManagerStatus.java [hunk 1]
```diff
@@ -144,10 +144,82 @@ private synchronized boolean isRunning() {
...
+                                                      final ReefServiceProtos.State to) {
+
+      // handle diagonal elements of the state transition matrix
+    if (from.equals(to)){
+      LOG.warning("Transition from " + from + " state to the same state. This shouldn't happen.");
+      return true;
+    }
+
+    // handle non-diagonal elements
+    switch (from.getNumber()) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/resourcemanager/ResourceManagerStatus.java]: The calls to `getNumber` shouldn't be needed in Java 7

**Comment by zerg-zerg** [comment]
@markusweimer  Updated.

**Comment by markusweimer** [comment]
I'll merge.

**Pull request closed without merging**